src/transformers/models/pix2struct/modeling_pix2struct.py

```diff
@@ -1759,54 +1738,29 @@ def forward(
             encoder_attentions=encoder_outputs.attentions,
         )
 
-    @torch.no_grad()
-    def generate(
+    def prepare_inputs_for_generation(
         self,
-        flattened_patches: torch.FloatTensor,
-        decoder_input_ids: Optional[torch.LongTensor] = None,
+        input_ids,
+        flattened_patches: Optional[torch.FloatTensor] = None,
         attention_mask: Optional[torch.FloatTensor] = None,
-        decoder_attention_mask: Optional[torch.LongTensor] = None,
-        **generate_kwargs,
+        decoder_attention_mask: Optional[torch.BoolTensor] = None,
+        past_key_values=None,
+        head_mask=None,
+        decoder_head_mask=None,
+        cross_attn_head_mask=None,
+        use_cache=None,
+        encoder_outputs=None,
+        **kwargs,
     ):
-        r"""
-        Returns:
-
-        Example:
-
-        ```python
-        >>> from PIL import Image
-        >>> import requests
-        >>> from transformers import AutoProcessor, Pix2StructForConditionalGeneration
-
-        >>> processor = AutoProcessor.from_pretrained("google/pix2struct-textcaps-base")
-        >>> model = Pix2StructForConditionalGeneration.from_pretrained("google/pix2struct-textcaps-base")
-
-        >>> conditional_text = "A stop sign"
-        >>> url = "https://www.ilankelman.org/stopsigns/australia.jpg"
-        >>> image = Image.open(requests.get(url, stream=True).raw)
-
-        >>> inputs = processor(images=image, text=conditional_text, return_tensors="pt", add_special_tokens=True)
-
-        >>> # forward pass
-        >>> outputs = model.generate(**inputs)
-        >>> print(processor.batch_decode(outputs, skip_special_tokens=True))
-        ['A stop sign the street with a sign that says yes']
-        ```"""
-        batch_size, _, _ = flattened_patches.shape
-
-        vision_outputs = self.encoder(flattened_patches=flattened_patches, attention_mask=attention_mask)
-
-        image_embeds = vision_outputs[0]
-
-        if isinstance(decoder_input_ids, torch.Tensor):
-            # check if the first element of `input_ids` is equal to `decoder_input_ids`:
-            if (decoder_input_ids[:, 0] != self.decoder_start_token_id).all().item():
-                # add `decoder_input_ids` as first token to `input_ids`
-                decoder_input_ids = torch.cat(
+        if isinstance(input_ids, torch.Tensor):
+            # check if the first element of `input_ids` is equal to `input_ids`:
+            if (input_ids[:, 0] != self.config.decoder_start_token_id).all().item():
```

Comment:
This `if` wouldn't be needed if the tokenizer had the BOS token defined 😢  It's too late to change that, correct (users may have cloned the model)?

If it's too late to retroactively fix this, let's add add a comment about why we need this `if` :)

Reply:
cc @younesbelkada since the model isn't available in a new PyPi release yet, we can probably update this.

Reply:
There is a short window where we can do breaking changes yes, until next week.

Reply:
This part still need updating no?

Reply:
Pinging @younesbelkada here